Clean publish and eject

Jiansen opened the pull request:
- [x]  separate dev config and prod config, including paths.js 
  - part of #191 
- [x]  cleanup code on eject
  - same as in #206 , can close 206 if this PR is accepted.
  - as discussed in #206, the `--smoke-test` tag is kept in production code for now
- [x] on release, only ship configurations for production code
  - in the node_module installed at users side, there will be no `--debug-template` flag
- [x] same as above, in `e2e` test, when testing cli, only pack production code
- [x] on local dev (i.e. build and start), use dev configuration

Disscussion 1:  add parameter `stage` to pack.sh so that
- pack.sh prod only ship prod configurations
- pack.sh dev only ship dev configurations
- pack.sh test only ship test configurations
- etc.

This can be done by removing `files` in package.json, and creating a few versions of .npmignore files (e.g. .npmignore.prod).

Disscussion 2: further split paths.prod.js to 2 files, paths.beforeEject.js and paths.afterEject.js; remove the `// Dead code on eject:start/end` block.

Question:
[start.js](https://github.com/facebookincubator/create-react-app/blob/gaearon-0.2.0-changelog/scripts/start.js#L16) always use the dev config for webpack.  Is there a reason for this?  In my local test, it seems fine to use production config in a generated app.

Jiansen commented:
Don't merge for now.

I found this in CI build console.  `e2e` isn't completed there.  Might suggest another issue to fix.
```
++++tasks/pack.sh
tasks/e2e.sh: line 50: tasks/pack.sh: Permission denied
```

vjeux commented:
You probably want to chmod +x tasks/pack.sh

Jiansen commented:
@vjeux ,
 
Yes, I checked the permission on my local machine, it has +x permission.  `git status` doesn't show anything to commit.  If permission change is not picked by `git`, I will probably delete and re-add the `pack.sh` file.

Thankfully, the last build (222.1) looks OK.

gaearon commented on config/webpack.config.dev.js:
```diff
@@ -11,7 +11,7 @@ var path = require('path');
 var webpack = require('webpack');
 var HtmlWebpackPlugin = require('html-webpack-plugin');
-var paths = require('./paths');
+var paths = require('./')('paths');
```
I’m worried that this convention is going to look super confusing to people after ejecting (or even for those who want to contribute to this project). What alternative solutions have you considered?

Jiansen replied:
I agree that the code is confusing to some extent.  In fact, I leant very recently that `require` looks for `index.js` when the required path is a folder.

Two ideas:

1.  replace it with `require('../config')('paths');` even if the file is in the same folder.  At least the word `config` appears in the code to suggest the meaning of this line.

2. create a *symbolic name* **config** so that the above code is rewritten as `require('config')('paths')`.

I don't know how to do 2 honestly.  I will take a look at [this method](https://gist.github.com/branneman/8048520#1-the-symlink) and search for others.

There is another decision worthy a discussion: should `config/index.js` export a function (like the current change) or an object with fields like `paths` and so on.

It is more conventional for a module to export an object or a constructor.  But exporting a function has advantages in this case:

1.  the code is short, and we don't need to introduce a new field every time when a new configuration is introduced; simply create a *.prod|dev.js file.

2. we could add an error message/exception when developers made a typo on importing a configuration, let the typo be exposed earlier.   

3. in test/dev, we are still able to import a prod config (e.g. the current `build.js` always uses `webpack.prod`)

gaearon commented:
Looking at this, I think we should keep it simple and not mess with the filenames.
If `paths.js` is the only thing that really needs branching, let’s just do something like

```js
// paths.js

module.exports = {
  // ejected paths
}

// @remove-on-eject-begin
if (...) {
  module.exports = // local dev paths
} else if (...) {
  module.exports = // dependency paths
}
// @remove-on-eject-end
```

Jiansen commented:
> If `paths.js` is the only thing that really needs branching  ...

Changes in this PR cut off code at 2 places
1. when `pack`/`publish` the project as a npm module
2. when a user runs `eject`

If we don't do 1, then dev-only flags like `--debug-template` are still available to users, though they need to dig a bit into the source code of this open source project :p

I like your syntax of `// @remove-on-eject-begin/end`.  I will refactor code in PR #206, which only does 2.  We could review if features in the top comment are needed or not, and update the code accordingly.

gaearon commented:
We can do (1) with separate flags.
Maybe something like:

```js
// paths.js

module.exports = {
  // ejected paths
}

// @remove-on-publish-begin
if (...) {
  module.exports = // local dev paths
}
// @remove-on-publish-end

// @remove-on-eject-begin
if (...) {
  module.exports = // dependency paths
}
// @remove-on-eject-end
```

Jiansen commented:
Yes, merging *.prod.js file and *.dev.js into one file is a good idea for code maintenance.

I will try later today or over this weekend.  Thanks for your advice.

Jiansen commented:
Refactored and rebased previous changes as suggested.  To summarise
- [X] tag dev-only code with // @remove-on-publish-begin/end tag
- [X] tag pre-eject-only code with // @remove-on-eject-begin/end tag

As per the support of the above 2 tags, if code looks OK, this PR is ready to be merged.

If we wanted to merge `babel.[prod|dev].js` and `webpack.config.[prod|dev].js` as well, I have 2 questions

1. `*.prod.js` contains some code that is not in `*.dev.js`
2. Currently `build.js` always uses `webpack.config.prod` whereas `start.js` always use `webpack.config.dev`

```
// `build.js`
var config = require('../config/webpack.config.prod');
```
```
// `start.js`
var config = require('../config/webpack.config.dev');
```

In an earlier test, it seems fine to let `start.js` uses `prod` config in production environment.  I haven't tested if `build.js` could use `dev` config in local development environment.

I will search for `webpack` related issues and PRs to understand why those two files are tied to specific configurations (runtime/build efficiency?).  If someone has a short answer, or knows the reference to a related issue/PR, please let me know.

gaearon commented on tasks/clean_pack.sh:
```diff
@@ -0,0 +1,42 @@
+# Exit the script on any command with non 0 return code
+# We assume that all the commands in the pipeline set their return code
+# properly and that we do not need to validate that the output is correct
+set -e
```
Can you please use the same error handling mechanism that e2e test [now uses](https://github.com/facebookincubator/create-react-app/blob/0e5abada75db99d3f8d3598afe013c21dc810e85/tasks/e2e.sh#L11-L35)?

gaearon commented on tasks/release.sh:
```diff
@@ -30,6 +30,22 @@ if [ -n "$(git status --porcelain)" ]; then
+mkdir .clean-pack
+
+# copy files to folder .clean-pack
+rsync -av --exclude='.git' --exclude='.clean-pack' --exclude='node_modules' --exclude='tasks' './' '.clean-pack'  >/dev/null
```
It would help if comment explained why these specific folders are excluded.

gaearon commented:
I left a few final comments, I think it’s good to go when they are addressed.
Also please rebase, it doesn’t merge now.

Jiansen commented:
@gaearon , I rebased as suggested.

Do you think we should use the same error handling mechanism in `release.sh` as well, or run `e2e` at the beginning,  or assume project will always properly tested before a project admin runs `release.sh`?

gaearon commented:
I think it's relatively safe to assume the admin knows what they're shipping, release is slow enough without another e2e test.

gaearon commented on tasks/handle_error_exit.sh:
```diff
@@ -0,0 +1,21 @@
+
+# print error messages to stderr
+# the cleanup function is optionally defined in caller script
+function handle_error {
```
I would prefer we keep these duplicated in all files. We’re not shell experts so it’s easier if we don’t `source` any files. Duplication is not a problem for us.

Jiansen replied:
will update

gaearon commented on scripts/eject.js:
```diff
@@ -75,6 +75,8 @@ prompt(
       // Remove license header from AppleScript
       .replace(/^--.*\n/gm, '')
+      // Remove dead code on eject
+      .replace(/\/\/ @remove-on-eject-begin([\s\S]*?)\/\/ @remove-on-eject-end/g, '')
```
Can we make it remove the license headers via the same mechanism?

Jiansen replied:
@gaearon ,

I want to double check your question.

The current RegExp `/^\/\*\*(\*(?!\/)|[^*])*\*\//` removes all multiline JS comments, which could contains useful information or other License.

We have 2 solutions for this problem

#### option 1
a.  Remove those 4 lines in `eject.js`

```
// Remove license header from JS
.replace(/^\/\*\*(\*(?!\/)|[^*])*\*\//, '')
// Remove license header from AppleScript
.replace(/^--.*\n/gm, '')
```
b. add a RegExp for using `@remove-on-eject-begin/end` in AppleScript.

c. add `@remove-on-eject-begin/end` before/after the license statement of each file.

#### option 2
Update the RegExp so that it only matches Facebook license header in JS and AppleScript


I prefer **option 1**, which gives us the potential of retaining license in some files (e.g util scripts, 3rd party code).

gaearon replied:
Yes, option 1 is what I meant.

vjeux replied:
> The current RegExp /^\/\*\*(\*(?!\/)|[^*])*\*\// removes all multiline JS comments, which could contains useful information or other License.

This is not correct. It only removes the first multiline comment (because it doesn't have /g at the end). So all the other multiline comments on the file will be preserved just fine.

Jiansen commented:
Updated code to address the issue. Rebased to current master.

Please let me know if some of the followings need to be addressed.
- LICENSE clean up tags are only added to those files copied on eject
- license is added to config/polyfills.js
- no license for `config/flow/*.js.flow`
- no license for files in the `template` folder
- license in applescript is replace using multiline comment style

Jiansen commented:
Rebased.

@gaearon , do you plan to add this PR back to milestone 0.3.0, if not merging it soon?

Currently, minor changes to config/paths.js will cause code conflicts due reordering of configs.  I don't mind to rebase again when it is a good time to deliver.  Just curious about the plan.

gaearon commented:
Let’s get this in and maybe iterate on it later.
Thanks.

gaearon commented:
<img width="513" alt="screen shot 2016-09-02 at 16 55 56" src="https://cloud.githubusercontent.com/assets/810438/18209955/2bc1ff7c-712e-11e6-847b-5e5eea27c506.png">

This is really nice. 👍 
If you’d like, please feel free to make a followup PR to make eject result more natural.
For example I think it might make sense to try to write `babel.dev.js` and `babel.prod.js` without all those crazy `require.resolve`s and then put the `require.resolve` mapping behind an `// @remove-on-eject`.

Jiansen commented:
@gaearon , good point.  I will try as you suggested.
